File: lib/ts-sdk/examples/mock-api-server.ts
/**
 * Minimal mock CommonGrants API server for running examples without the FastAPI template.
 *
 * Serves GET/POST routes that match the SDK client. Start with:
 *   pnpm example:server
 *
 * Then run any example (get, list, search, get-custom-fields) against http://localhost:8000.
 *
 * The search route understands `filters`: it applies the standard `status`
 * filter, and echoes every received filter (including `customFilters`) back in
 * `filterInfo.filters`, so examples can show that registered and ad-hoc custom
 * filters reached the server. Import `createMockServer()` to run it in-process.
 */

import { createServer, type Server } from "http";

const PORT = parseInt(process.env.PORT ?? "8000", 10);

// =============================================================================
// Mock Opportunities
// =============================================================================

const MOCK_OPPORTUNITIES = [
  {
    id: "573525f2-8e15-4405-83fb-e6523511d893",
    title: "STEM Education Grant Program",
    description: "A grant program focused on improving STEM education.",
    status: { value: "open" },
    createdAt: "2025-01-01T00:00:00Z",
    lastModifiedAt: "2025-01-15T00:00:00Z",
    customFields: {
      legacyId: {
        name: "legacyId",
        fieldType: "integer",
        value: 12345,
        description: "Legacy system opportunity ID",
      },
      programCode: {
        name: "programCode",
        fieldType: "string",
        value: "STEM-ED",
        description: "Funding program code",
      },
    },
  },
  {
    id: "a1b2c3d4-e5f6-7890-abcd-ef1234567890",
    title: "Community Development Grant",
    description: "Funding for community development projects.",
    status: { value: "open" },
    createdAt: "2025-01-02T00:00:00Z",
    lastModifiedAt: "2025-01-16T00:00:00Z",
    customFields: {
      legacyId: {
        name: "legacyId",
        fieldType: "integer",
        value: 12346,
        description: "Legacy system opportunity ID",
      },
      programCode: {
        name: "programCode",
        fieldType: "string",
        value: "COMM-DEV",
        description: "Funding program code",
      },
    },
  },
  {
    id: "b2c3d4e5-f6a7-8901-bcde-f12345678901",
    title: "Education Initiative",
    description: "Support for education initiatives.",
    status: { value: "forecasted" },
    createdAt: "2025-01-03T00:00:00Z",
    lastModifiedAt: "2025-01-17T00:00:00Z",
    customFields: {
      legacyId: {
        name: "legacyId",
        fieldType: "integer",
        value: 12347,
        description: "Legacy system opportunity ID",
      },
      programCode: {
        name: "programCode",
        fieldType: "string",
        value: "EDU-INIT",
        description: "Funding program code",
      },
    },
  },
];

// =============================================================================
// Helper Functions
// =============================================================================

function send(res: import("http").ServerResponse, status: number, body: unknown) {
  res.writeHead(status, { "Content-Type": "application/json" });
  res.end(JSON.stringify(body));
}

function parseQuery(url: string): Record<string, string> {
  const i = url.indexOf("?");
  if (i === -1) return {};
  const out: Record<string, string> = {};
  for (const part of url.slice(i + 1).split("&")) {
    const [k, v] = part.split("=").map(decodeURIComponent);
    if (k && v !== undefined) out[k] = v;
  }
  return out;
}

/** Status values carried by an opportunity, for the standard `status` filter. */
function statusOf(opp: (typeof MOCK_OPPORTUNITIES)[number]): string {
  return opp.status.value;
}

// =============================================================================
// Mock API Server
// =============================================================================

/** Builds the mock server without starting it, so examples can run it in-process. */
export function createMockServer(): Server {
  return createServer((req, res) => {
    const url = req.url ?? "";
    const method = req.method ?? "GET";
    const path = url.split("?")[0];

    // =========================================================================
    // GET /common-grants/opportunities
    // =========================================================================

    if (method === "GET" && path === "/common-grants/opportunities") {
      const q = parseQuery(url);
      const page = Math.max(1, parseInt(q.page ?? "1", 10));
      const pageSize = Math.min(100, Math.max(1, parseInt(q.pageSize ?? "25", 10)));
      const start = (page - 1) * pageSize;
      const items = MOCK_OPPORTUNITIES.slice(start, start + pageSize);
      send(res, 200, {
        status: 200,
        message: "Success",
        items,
        paginationInfo: {
          page,
          pageSize,
          totalItems: MOCK_OPPORTUNITIES.length,
          totalPages: Math.ceil(MOCK_OPPORTUNITIES.length / pageSize),
        },
      });
      return;
    }

    // =========================================================================
    // GET /common-grants/opportunities/:id
    // =========================================================================

    if (method === "GET" && path.startsWith("/common-grants/opportunities/")) {
      const id = path.slice("/common-grants/opportunities/".length);
      const opp = MOCK_OPPORTUNITIES.find(o => o.id === id) ?? {
        id,
        title: "Mock Opportunity",
        description: "Mock opportunity for examples.",
        status: { value: "open" },
        createdAt: "2025-01-01T00:00:00Z",
        lastModifiedAt: "2025-01-01T00:00:00Z",
      };
      send(res, 200, {
        status: 200,
        message: "Success",
        data: opp,
      });
      return;
    }

    // =========================================================================
    // POST /common-grants/opportunities/search
    // =========================================================================

    if (method === "POST" && path === "/common-grants/opportunities/search") {
      let body = "";
      req.on("data", chunk => {
        body += chunk;
      });
      req.on("end", () => {
        let items = [...MOCK_OPPORTUNITIES];
        let filters: Record<string, unknown> = {};
        try {
          const parsed = body
            ? (JSON.parse(body) as { search?: string; filters?: Record<string, unknown> })
            : {};
          if (parsed.search && typeof parsed.search === "string") {
            const q = parsed.search.toLowerCase();
            items = items.filter(o => o.title.toLowerCase().includes(q));
          }
          filters = parsed.filters ?? {};
          // Apply the standard `status` filter; custom filters are echoed but
          // not applied by this mock.
          const status = filters.status as { value?: unknown } | undefined;
          if (Array.isArray(status?.value)) {
            const allowed = new Set(status.value as string[]);
            items = items.filter(o => allowed.has(statusOf(o)));
          }
        } catch {
          // ignore malformed bodies
        }
        send(res, 200, {
          status: 200,
          message: "Success",
          items,
          paginationInfo: {
            page: 1,
            pageSize: items.length > 0 ? items.length : 1,
            totalItems: items.length,
            totalPages: 1,
          },
          sortInfo: { sortBy: "lastModifiedAt", sortOrder: "desc" },
          // Echo the received filters so a consumer can confirm registered and
          // ad-hoc custom filters reached the server.
          filterInfo: { filters },
        });
      });
      return;
    }

    send(res, 404, { status: 404, message: "Not found" });
  });
}

// =============================================================================
// Start Server (only when run directly, e.g. `pnpm example:server`)
// =============================================================================
if (process.argv[1]?.includes("mock-api-server")) {
  createMockServer().listen(PORT, () => {
    console.log(`Mock CommonGrants API listening on http://localhost:${PORT}`);
    console.log("");
    console.log("Run examples in another terminal:");
    console.log("  pnpm example:list");
    console.log("  pnpm example:get 573525f2-8e15-4405-83fb-e6523511d893");
    console.log("  pnpm example:search education");
    console.log("  pnpm example:get-custom-fields 573525f2-8e15-4405-83fb-e6523511d893");
    console.log("");
  });
}

File: lib/python-sdk/examples/mock_api_server.py
#!/usr/bin/env python3
"""Minimal mock CommonGrants API for running the examples without a real server.

Serves the GET and POST routes the SDK client calls. The search route applies the
standard ``status`` filter so the happy path returns a realistic subset. Start it
in one terminal:

    poetry run python examples/mock_api_server.py

Then run an example against http://localhost:8000 in another terminal, and stop
the server with Ctrl-C when you are done.
"""

from __future__ import annotations

import json
import os
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import Any

PORT = int(os.environ.get("PORT", "8000"))

# Two opportunities, each carrying a "programCode" custom field so an example can
# show a typed custom field on a parsed row.
MOCK_OPPORTUNITIES: list[dict[str, Any]] = [
    {
        "id": "573525f2-8e15-4405-83fb-e6523511d893",
        "title": "STEM Education Grant Program",
        "description": "A grant program focused on improving STEM education.",
        "status": {"value": "open", "description": "Open"},
        "createdAt": "2025-01-01T00:00:00Z",
        "lastModifiedAt": "2025-01-15T00:00:00Z",
        "customFields": {
            "programCode": {
                "name": "programCode",
                "fieldType": "string",
                "value": "STEM-ED",
            },
        },
    },
    {
        "id": "a1b2c3d4-e5f6-7890-abcd-ef1234567890",
        "title": "Community Development Grant",
        "description": "Funding for community development projects.",
        "status": {"value": "forecasted", "description": "Forecasted"},
        "createdAt": "2025-01-02T00:00:00Z",
        "lastModifiedAt": "2025-01-16T00:00:00Z",
        "customFields": {
            "programCode": {
                "name": "programCode",
                "fieldType": "string",
                "value": "COMM-DEV",
            },
        },
    },
]


def _requested_statuses(filters: dict[str, Any]) -> list[str] | None:
    """The status values a search asked for, if it included a status filter."""
    status = filters.get("status")
    if isinstance(status, dict) and isinstance(status.get("value"), list):
        return [str(v) for v in status["value"]]
    return None


class Handler(BaseHTTPRequestHandler):
    def log_message(self, format: str, *args: Any) -> None:
        """Silence the default per-request logging."""

    def _send(self, status: int, body: dict[str, Any]) -> None:
        payload = json.dumps(body).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)

    def do_GET(self) -> None:
        path = self.path.split("?")[0]
        if path == "/common-grants/opportunities":
            self._send(
                200,
                {
                    "status": 200,
                    "message": "Success",
                    "items": MOCK_OPPORTUNITIES,
                    "paginationInfo": {
                        "page": 1,
                        "pageSize": 25,
                        "totalItems": len(MOCK_OPPORTUNITIES),
                        "totalPages": 1,
                    },
                },
            )
            return
        if path.startswith("/common-grants/opportunities/"):
            opp_id = path.rsplit("/", 1)[-1]
            opp = next(
                (o for o in MOCK_OPPORTUNITIES if o["id"] == opp_id),
                MOCK_OPPORTUNITIES[0],
            )
            self._send(200, {"status": 200, "message": "Success", "data": opp})
            return
        self._send(404, {"status": 404, "message": "Not found"})

    def do_POST(self) -> None:
        if self.path.split("?")[0] != "/common-grants/opportunities/search":
            self._send(404, {"status": 404, "message": "Not found"})
            return

        length = int(self.headers.get("Content-Length", "0") or "0")
        raw = self.rfile.read(length) if length else b""
        try:
            body = json.loads(raw) if raw else {}
        except json.JSONDecodeError:
            body = {}

        filters = body.get("filters") or {}
        items = list(MOCK_OPPORTUNITIES)
        wanted = _requested_statuses(filters) if isinstance(filters, dict) else None
        if wanted is not None:
            items = [o for o in items if o["status"]["value"] in wanted]

        self._send(
            200,
            {
                "status": 200,
                "message": "Success",
                "items": items,
                "paginationInfo": {
                    "page": 1,
                    "pageSize": max(len(items), 1),
                    "totalItems": len(items),
                    "totalPages": 1,
                },
                "sortInfo": {
                    "sortBy": "lastModifiedAt",
                    "sortOrder": "desc",
                    "customSortBy": None,
                    "errors": [],
                },
                "filterInfo": {"filters": {}, "errors": []},
            },
        )


def main() -> None:
    server = HTTPServer(("localhost", PORT), Handler)
    print(f"Mock CommonGrants API listening on http://localhost:{PORT}")
    print("Run an example in another terminal. Ctrl-C to stop.")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        server.server_close()


if __name__ == "__main__":
    main()
